[WIP][mobilidade-api] Pipeline de atualização do banco QRCode

Hellcassius opened the pull request:
Descrição:
Flow de atualização do banco postgres da API da mobilidade. O flow realiza as seguintes tarefas:
- Baixa e extrai arquivos de GTFS de uma pasta do GCS (definido como `gtfs/`, na raiz do bucket)
-  Concatena cada par de tabelas, SPPO e BRT, para gerar um "gtfs único", contendo toda a informação necessária para ser consumida pela API
-  Trata os dados para lidar com restrições de `primary key`, `foreign key`, restrições específicas do Django (nomes de colunas, nomes de tabelas) e considerações para o bom funcionamento do app (conversão de trip id's)
-  Carrega os novos dados via `psycopg2`, primeiramente removendo todos os dados presentes nas tabelas de gtfs (não altera tabelas como `feedback`) e então inserindo os dados tratados.
Essa versão para staging está sendo enviada com a conexão em um banco alternativo para testes, que está exposto para acesso externo, após testes finais, mudaremos a conexão para o banco de produção da API.

TODO:
- [ ] mudar secret_path e conteudos do secret para conectar no banco de produção

Hellcassius committed:
first commit

yxuo committed:
fix: working from extract_gtfs to execute_update

Hellcassius committed:
minor fixes, add logs, simplify paths

mergify[bot] committed:
Merge branch 'master' into staging/smtr-postgres-update

Hellcassius committed:
add to poetry.lock and pyproject.toml

Hellcassius committed:
merge master local

Hellcassius committed:
add psycopg2-binary, remove psycopg2

fernandascovino commented on pipelines/rj_smtr/constants.py:
```diff
@@ -129,6 +129,75 @@ class constants(Enum):  # pylint: disable=c0103
+        },
+    }
+
+    GTFS_RENAME_COLUMNS = {
```
isso é bem especifico do django, nao acho `GTFS_RENAME_COLUMNS` um bom nome

yxuo replied:
Renomeado para `GTFS_DJANGO_COLUMNS`, indicando que são colunas GTFS do Django.

fernandascovino commented on pipelines/rj_smtr/constants.py:
```diff
@@ -129,6 +129,75 @@ class constants(Enum):  # pylint: disable=c0103
+        "trips": {"route_id": "route_id_id"},
+    }
+
+    GTFS_RENAME_TABLES = {
```
mesmo comentário anterior

yxuo replied:
Renomeado para `GTFS_DJANGO_TABLES`.

fernandascovino commented on pipelines/rj_smtr/mobilidade_api/tasks.py:
```diff
@@ -0,0 +1,209 @@
+        trip = trips[trips.trip_id == trip_id].iloc[0]
+        return f"{trip.trip_short_name}{trip.direction_id}{trip.service_id}"
+
+    trip_id_dict = dict(zip(trips["trip_id"], trips["trip_id"].apply(rename_trip_id)))
```
por que tem essa operação toda? é só aplicar nas colunas da tabela, não é necessário uma função extra. ta complexo dms para uma tarefa mt simples

yxuo replied:
Essa funcionalidade foi removida, pois o Django já filtra as trips no backend.

Futuramente, ou caso haja um entendimento que seja necessário o Django **não** filtrar trips redundantes, mas usar o GTFS já filtrado, será preciso mapear as consequências de deixar os trip_id iguais.
Um dos impactos, por exemplo, é ter que remover a restrição de trip_id único na tabela frequencies ([gtfs.org](https://gtfs.org/schedule/reference/#:~:text=primary%20key%20(trip_id%2C%20start_time)))

yxuo committed:
sort folders, restart identity,dont rename trip_id

- Before concat, sort GTFS folders by name order or ascending size
- Rename constants to be more coherent
- Restart identity on truncate
- Dont rename trip_id for now
  Reason: DJango will have to remove GTFS unique constraint in
  frequencies.
  And to do this, is necessary to check if it breaks GTFS logic
  beyond just removing constraints.

Refs #410

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

mergify[bot] committed:
Merge branch 'master' into staging/smtr-postgres-update

Hellcassius committed:
simplify file check, better data type enforcing

Hellcassius committed:
Merge branch 'staging/smtr-postgres-update' of https://github.com/prefeitura-rio/pipelines into staging/smtr-postgres-update

Hellcassius committed:
add code owner

mergify[bot] committed:
Merge branch 'master' into staging/smtr-postgres-update

Hellcassius committed:
change labels to dev

Hellcassius committed:
Merge branch 'staging/smtr-postgres-update' of https://github.com/prefeitura-rio/pipelines into staging/smtr-postgres-update

mergify[bot] committed:
Merge branch 'master' into staging/smtr-postgres-update

mergify[bot] committed:
Merge branch 'master' into staging/smtr-postgres-update

yxuo commented on pipelines/rj_smtr/mobilidade_api/tasks.py:
```diff
@@ -0,0 +1,217 @@
+        dict: A dict containing {<table name>: <absolute path>}
+    """
+    table_paths = {}
+    # Remove SN routes from table `routes`
```
TODO: remover validação, não é mais necessáro

yxuo commented on pipelines/rj_smtr/mobilidade_api/tasks.py:
```diff
@@ -0,0 +1,217 @@
+        ~tables["routes"].route_short_name.astype(str).str.contains("SN")
+    ].copy()
+
+    # Cross checks to remove invalid foreign keys
```
TODO: Remover isso, não é necessário aqui

yxuo commented on pipelines/rj_smtr/mobilidade_api/tasks.py:
```diff
@@ -0,0 +1,217 @@
+                    dtype
+                )
+
+        # drop primary key duplicates
```
Não precisa mais remover duplicatas, agora usamos 1 tabela

yxuo commented on pipelines/rj_smtr/mobilidade_api/tasks.py:
```diff
@@ -0,0 +1,217 @@
+
+
+@task
+def treat_gtfs_tables(tables: Dict[str, pd.DataFrame]) -> dict:
```
renomear para pre_treatment_mobilidade_api_gtfs()

yxuo committed:
refactor: simplificar pré tratamento

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

mergify[bot] committed:
Merge branch 'master' into staging/smtr-postgres-update

mergify[bot] committed:
Merge branch 'master' into staging/smtr-postgres-update

mergify[bot] committed:
Merge branch 'master' into staging/smtr-postgres-update

mergify[bot] committed:
Merge branch 'master' into staging/smtr-postgres-update

mergify[bot] committed:
Merge branch 'master' into staging/smtr-postgres-update

mergify[bot] committed:
Merge branch 'master' into staging/smtr-postgres-update
